google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/SeekableByteChannelPrefetcher.java

```diff
@@ -0,0 +1,496 @@
+   */
+  public static SeekableByteChannel addPrefetcher(int bufferSizeMB, SeekableByteChannel channel) throws IOException {
+    return new SeekableByteChannelPrefetcher(channel, bufferSizeMB * 1024 * 1024);
+  }
+
+  /**
+   * WorkUnit holds a buffer and the instructions for what to put in it.
+   *
+   * <p>Use it like this:
+   * <ol>
+   *   <li> call()
+   *   <li> the data is now in buf, you can access it directly
+   *   <li> if need more, call resetForIndex(...) and go back to the top.
+   *   <li> else, call close()
+   * </ol>
+   */
+  private static class WorkUnit implements Callable<ByteBuffer>, Closeable {
+    public final ByteBuffer buf;
+    public long blockIndex;
+    private final SeekableByteChannel chan;
+    private final int blockSize;
+    private Future<ByteBuffer> futureBuf;
+
+    public WorkUnit(SeekableByteChannel chan, int blockSize, long blockIndex) {
+      this.chan = chan;
+      this.buf = ByteBuffer.allocate(blockSize);
+      this.futureBuf = null;
+      this.blockSize = blockSize;
+      this.blockIndex = blockIndex;
+    }
+
+    @Override
+    public ByteBuffer call() throws IOException {
+      long pos = ((long)blockSize) * blockIndex;
+      if (pos > chan.size()) {
+        return null;
+      }
+      if (pos < 0) {
+        // This should never happen, if the code's correct.
+        throw new IllegalArgumentException("blockIndex " + blockIndex + " has position " + pos + ": negative position is not valid.");
+      }
+      chan.position(pos);
+      // read until buffer is full, or EOF
+      while (chan.read(buf) >= 0 && buf.hasRemaining()) {}
+      return buf;
+    }
+
+    public ByteBuffer getBuf() throws ExecutionException, InterruptedException {
+      return futureBuf.get();
+    }
+
+    public WorkUnit resetForIndex(long blockIndex) {
+      this.blockIndex = blockIndex;
+      buf.clear();
+      futureBuf = null;
+      return this;
+    }
+
+
+    @Override
+    public void close() throws IOException {
+      chan.close();
+    }
+  }
+
+  /**
+   * Wraps the provided SeekableByteChannel within a SeekableByteChannelPrefetcher,
+   * using the provided buffer size.
+   *
+   * @param bufSize buffer size in bytes
+   * @param chan channel to wrap in the prefetcher
+   */
+  public SeekableByteChannelPrefetcher(SeekableByteChannel chan, int bufSize) throws IOException {
+    if (chan instanceof SeekableByteChannelPrefetcher) {
+      throw new IllegalArgumentException("Cannot put two prefetchers on the same channel.");
+    }
+    if (!chan.isOpen()) {
+      throw new IllegalArgumentException("channel must be open");
+    }
+    this.chan = chan;
+    if (bufSize <= 0) {
+      throw new IllegalArgumentException("bufSize must be positive");
+    }
+    this.size = chan.size();
+    if (bufSize > this.size) {
+      this.bufSize = (int)this.size;
+    } else {
+      this.bufSize = bufSize;
+    }
+    this.open = true;
+    int prefetcherIndex = prefetcherCount++;
```

Comment:
Yes, when debugging. Since this is a rarely-used feature, I didn't want to pay the price of an AtomicInteger. If the name's doubled in some debugging session it's not the end of the world.